### PR TITLE
GODRIVER-1199 Add UnmarshalJSON and MarshalJSON for primitive.Decimal128

### PR DESCRIPTION
**Pull request by iwysiu**

### CODE REVIEW

File: bson/primitive/decimal_test.go [hunk 1]
```diff
@@ -144,3 +148,42 @@ func TestParseDecimal128(t *testing.T) {
...
+		bytes, err := json.Marshal(decimal)
+		assert.Nil(t, err, "json.Marshal error: %v", err)
+		got := NewDecimal128(0, 0)
+		err = json.Unmarshal([]byte(bytes), &got)
+		assert.Nil(t, err, "json.Unmarshal error: %v", err)
+		assert.Equal(t, decimal.h, got.h, "expected h: %v got: %v", decimal.h, got.h)
+		assert.Equal(t, decimal.l, got.l, "expected l: %v got: %v", decimal.l, got.l)
+	})
+	t.Run("Unmarshal", func(t *testing.T) {
+		cases := append(bigIntTestCases,
```
Comment (kevinAlbs) [inline bson/primitive/decimal_test.go:164]: Ooh, good idea reusing these existing test cases.

File: bson/primitive/decimal.go [hunk 2]
```diff
@@ -211,6 +212,22 @@ func (d Decimal128) IsZero() bool {
 	return d.h == 0 && d.l == 0
 }
 
+// MarshalJSON returns Decimal128 as a string.
+func (d Decimal128) MarshalJSON() ([]byte, error) {
+	return json.Marshal(d.String())
+}
+
+// UnmarshalJSON creates a primitive.DateTime from a JSON string.
```
Comment (kevinAlbs) [inline bson/primitive/decimal.go:220]:
```suggestion
// UnmarshalJSON creates a primitive.Decimal128 from a JSON string.
```

File: bson/primitive/decimal.go [hunk 3]
```diff
@@ -211,6 +212,22 @@ func (d Decimal128) IsZero() bool {
 	return d.h == 0 && d.l == 0
 }
 
+// MarshalJSON returns Decimal128 as a string.
```
Comment (kevinAlbs) [inline bson/primitive/decimal.go:215]: Perhaps best discussed offline, but I have a high level question.

Is the main use case for marshaling Decimal128 into a string (as opposed to something else) for readability?

Could we encode it as the canonical extended JSON form: `{"$numberDecimal": "1.23"}`? It's more verbose, but also less ambiguous.

It seems like this is consistent with other BSON types (e.g. ObjectID marshals into a string).

File: bson/primitive/decimal.go [hunk 4]
```diff
@@ -211,6 +212,22 @@ func (d Decimal128) IsZero() bool {
...
+func (d Decimal128) MarshalJSON() ([]byte, error) {
+	return json.Marshal(d.String())
+}
+
+// UnmarshalJSON creates a primitive.DateTime from a JSON string.
+func (d *Decimal128) UnmarshalJSON(b []byte) error {
```
Comment (divjotarora) [inline bson/primitive/decimal.go:221]: As discussed offline, this should also handle the extended JSON format as well as the format generated by `json.Marshal` before this work was done.

File: bson/primitive/decimal.go (hunk 3 above)
Comment (iwysiu) [reply bson/primitive/decimal.go:215]: Not sure, but I think we discussed this offline? But the idea is users don't expect the extended json syntax, and we want to keep consistency.

File: bson/primitive/decimal.go (hunk 4 above)
Comment (iwysiu) [reply bson/primitive/decimal.go:221]: done. Added Extended JSON support, the format generated before this didn't store anything to unmarshal.

**Review by benjirewis — APPROVED** [review]
LGTM 🎉

File: bson/primitive/decimal.go [hunk 5]
```diff
@@ -211,6 +212,40 @@ func (d Decimal128) IsZero() bool {
 	return d.h == 0 && d.l == 0
 }
 
+// MarshalJSON returns Decimal128 as a string.
+func (d Decimal128) MarshalJSON() ([]byte, error) {
+	return json.Marshal(d.String())
+}
+
+// UnmarshalJSON creates a primitive.Decimal128 from a JSON string.
+func (d *Decimal128) UnmarshalJSON(b []byte) error {
```
Comment (divjotarora) [inline bson/primitive/decimal.go:221]: Seems like it's a common pattern to check for the string `null` so that JSON null values can be unmarshalled to non-pointer data types. The standard library [time.Time](https://golang.org/src/time/time.go?s=38690:38737#L1246) type does this and we mimic that behavior for [primitive.DateTime](https://github.com/mongodb/mongo-go-driver/blob/60f76f5b1810553090503950fce37c876253217b/bson/primitive/primitive.go#L52).

Comment (iwysiu) [reply bson/primitive/decimal.go:221]: done

File: bson/primitive/decimal.go (hunk 3 above)
Comment (kevinAlbs) [reply bson/primitive/decimal.go:215]: Yes, from what I gathered we don't recommend users marshal directly to JSON instead of extended JSON, but they do and expect it to work.

File: bson/primitive/decimal.go [hunk 6]
```diff
@@ -211,6 +212,47 @@ func (d Decimal128) IsZero() bool {
...
+	if err != nil {
+		return err
+	}
+	str, ok := res.(string)
+
+	// Extended JSON
+	if !ok {
+		m, ok := res.(map[string]interface{})
+		if !ok {
+			return errors.New("not an extended JSON Decimal128")
```
Comment (kevinAlbs) [inline bson/primitive/decimal.go:240]: Nit: Can we add the reason to the message to make it easier to debug unmarshaling an invalid Decimal128? Like:
`not an  extended JSON Decimal128: expected document"`

File: bson/primitive/decimal_test.go [hunk 7]
```diff
@@ -160,15 +160,34 @@ func TestDecimal128_JSON(t *testing.T) {
...
+	})
+	t.Run("unmarshal", func(t *testing.T) {
 		cases := append(bigIntTestCases,
 			[]bigIntTestCase{
 				{s: "-0001231.453454000000565600000000E-21", h: 0xafe6000003faa269, l: 0x81cfeceaabdb1800},
 				{s: "12345E+21", h: 0x306a000000000000, l: 12345},
 				{s: "0.10000000000000000000000000000000000000000001", remark: "parse fail"},
 				{s: ".125e1", h: 0x303c000000000000, l: 125},
 				{s: ".125", h: 0x303a000000000000, l: 125},
-				{s: "null", remark: "parse fail"},
```
Comment (kevinAlbs) [inline bson/primitive/decimal_test.go]: Can we add a test case to ensure unmarshaling the string "null" results in an empty Decimal128, but no error?

**Review by kevinAlbs — APPROVED** [review]
LGTM with a nit and additional test case!

File: bson/primitive/decimal_test.go (hunk 7 above)
Comment (iwysiu) [reply bson/primitive/decimal_test.go]: yup! added on line 173. These test cases put quotes around the string fed in, so I couldn't do it in this block

File: bson/primitive/decimal.go (hunk 6 above)
Comment (iwysiu) [reply bson/primitive/decimal.go]: done

File: bson/primitive/decimal_test.go [hunk 8]
```diff
@@ -144,3 +148,61 @@ func TestParseDecimal128(t *testing.T) {
...
+
+		got := NewDecimal128(0, 0)
+		err := json.Unmarshal([]byte(extJSON), &got)
+		assert.Nil(t, err, "json.Unmarshal error: %v", err)
+		assert.Equal(t, want.h, got.h, "expected h: %v got: %v", want.h, got.h)
+		assert.Equal(t, want.l, got.l, "expected l: %v got: %v", want.l, got.l)
+	})
+	t.Run("unmarshal null", func(t *testing.T) {
+		want := NewDecimal128(0, 0)
+		extJSON := `{"$numberDecimal": null}`
```
Comment (divjotarora) [inline bson/primitive/decimal_test.go:175]: Is this the behavior we want? My suggestion for handling `null` was just to handle document fields like `{a: "null"}`, not `{a: {$numberDecimal: "null"}}`.The special handling for "null" comes from a push in the Go community to allow `null` values to be unmarshalled into non-pointer struct fields. This gains convenience when communicating across languages and API boundaries at the expense of type safety and roundtrip-ability. However, IMO, if we get as far as having a `$numberDecimal` field, we should expect it to be a proper extjson decimal 128 string.

Comment (iwysiu) [reply bson/primitive/decimal_test.go]: Agreed. That's what I meant for this test to be, and it turns out that `{"$numberDecimal": null}` correctly fails to unmarshal

File: bson/primitive/decimal.go [hunk 9]
```diff
@@ -211,6 +212,47 @@ func (d Decimal128) IsZero() bool {
 	return d.h == 0 && d.l == 0
 }
 
+// MarshalJSON returns Decimal128 as a string.
+func (d Decimal128) MarshalJSON() ([]byte, error) {
+	return json.Marshal(d.String())
+}
+
+// UnmarshalJSON creates a primitive.Decimal128 from a JSON string.
```
Comment (divjotarora) [inline bson/primitive/decimal.go:220]: This no longer supports just a string. We should document the supported formats and their behaviors (i.e. the string that would be outputted by `MarshalJSON`, extjson $numberDecimal values, and the string `null`).

File: bson/primitive/decimal_test.go [hunk 10]
```diff
@@ -144,3 +148,61 @@ func TestParseDecimal128(t *testing.T) {
...
+			}...)
+		for _, c := range cases {
+			input := fmt.Sprintf(`{"foo": %q}`, c.s)
+			var got map[string]Decimal128
+			err := json.Unmarshal([]byte(input), &got)
+
+			switch c.remark {
+			case "overflow", "parse fail":
+				assert.NotNil(t, err, "expected Unmarshal error, got nil")
+			case "", "rounding", "subnormal", "clamped", "NaN", "Infinity", "-Infinity":
```
Comment (divjotarora) [inline bson/primitive/decimal_test.go:200]: nit: Instead of listing out all of the possible values here, should we just make this `default`?

**Review by divjotarora — APPROVED** [review]
Very minor comments, but the code LGTM. No need to request another review from me once they're addressed.
